先复习一下事务的定义：事务是一个由多个操作组成的工作单元，相当于绑定了一系列增删改操作，这些操作要么全部成功，要么全部回滚，以保证数据的一致性。事务具有四个关键属性：原子性、一致性、隔离性和持久性（`ACID`）。

开启一个事务，获取带事务的数据库连接对象，一般命名为`tx`：

```go
tx := db.Begin()
```

`tx`也是一个`*gorm.DB`类型的对象，被称为事务对象，后续用它来执行数据库操作。

首先执行一系列增删改操作，如下所示：

```go
tx.Create(&YourModel{...})
tx.Model(&YourModel{}).Where("condition").Update("column", value)
tx.Where("condition").Delete(&YourModel{})
```

如果所有的操作都能成功执行，无`error`，提交事务：

```go
tx.Commit()
```

如果事务中任何操作失败，回滚事务：

```go
tx.Rollback()
```

事务提交和回滚的示例如下：

```go
// 这里模拟数据库操作出现错误，进行事务回滚的情况
err := errors.New("数据库错误")
if err != nil {
	tx.Rollback()
    return nil, err
}
tx.Commit()
```

还有一种操作事务的方法。`Gorm`有一个方法`Transaction`，它的方法签名是这样的：

```go
func (db *DB) Transaction(fc func(tx *DB) error, opts ...*sql.TxOptions) (err error)
```

参数中的`fc`就是封装要执行的一系列增删改操作的函数，根据返回的`error`是否为`nil`判断事务提交还是回滚。

使用这个方法，事务操作代码如下所示：

```go
err := db.Transaction(func(tx *gorm.DB) error {
	err := tx.Model(&dao.UserInfo{}).Where("user_name = ?", "zhangsan").Update("age", 3).Error
	if err != nil {
		return err
	}
	err = tx.Model(&dao.UserInfo{}).Where("user_name = ?", "Bob").Update("age", 4).Error
	if err != nil {
		return err
	}
	return nil
})

if err != nil {
	fmt.Println("transaction failed!")
} else {
	fmt.Println("transaction success!")
}
```

最外层返回的`err`会是内层代码最先返回的错误，只要最外层拿到了`err`，说明事务已被回滚。

推荐使用调用`Transaction`方法的方式来处理事务操作。当发生错误时，只需在内部函数中返回`error`，事务就会自动回滚。这样能够有效避免开发人员在分支逻辑中遗漏调用`tx.Rollback()`，从而导致事务未能正确回滚的风险。

如果事务发生回滚，那么代码中所有通过该`tx`对象所执行的增、删、改操作都会被一并回滚。这些操作在事务执行期间仅存在于事务上下文中，如果事务最终`COMMIT`，数据库会把事务期间的所有修改正式写入数据文件并持久化；如果事务最终`ROLLBACK`，数据库会丢弃事务期间的所有修改，相当于这些操作从未发生过。
